docs/debugging.rst

```diff
@@ -0,0 +1,249 @@
+
+
+When your debugging work is done, simply set ``debug=False``. Now ``assert`` will be ignored
+therefore no runtime overhead, making your program easy to debug without losing performance.
+
+
+Compile-time ``ti.static_assert``
+---------------------------------
+
+Like ``ti.static_print``, we also provide a static version of ``assert``:
+``ti.static_assert``, it can be useful to make assertion on data type / dimention / shape.
+It works no matter whether ``debug=True`` is specified. When assertion fails, it will
+raise ``AssertionError`` as a Python-scope ``assert`` does.
+
+For example:
+
+.. code-block:: python
+
+    @ti.func
+    def is_odd(x: ti.template()):
+        ti.static_assert(x.data_type() == ti.i32, "is_odd() is only supported for i32")
+        return x % 2 == 1
+
+
+Tips for debugging
+------------------
+
+Debugging a Taichi program can be hard even with the builtin tools above.
+Taichi developers are currently devoting themselves in improving error messages and warnings
+to help user find potential BUGs in their programs.
+
+Here we collected some common BUGs that one might encounter with a Taichi program:
+
+Static typing system
+++++++++++++++++++++
+
+Taichi pretends that it's a dynamical-typed language like Python, but it's actually a
+statically-typed language which will be translated into high performance CPU/GPU instructions.
+
+So the code behavior in Taichi-scope is actually very different from Python-scope!
+
+Type of a variable is simply **determined at its first initialization and never changes later**.
+
+Although static-type provides better performance and simplicity, but may leads to BUGs if
+users not distinguished Taichi-scope from Python-scope, e.g.:
+
+.. code-block:: python
+
+    @ti.kernel
+    def buggy():
+        ret = 0  # 0 is a integer, so `ret` is typed as int32
+        for i in range(3):
+            ret += 0.1 * i  # i32 += f32, the result is still stored in int32!
+        print(ret)  # will shows 0
+
+    buggy()
+
+The code above shows a common BUG due to the limitation of the static-type system.
+The Taichi compiler should shows a warning like:
+
+.. code-block:: none
+
+    [W 06/27/20 21:43:51.853] [type_check.cpp:visit@66] [$19] Atomic add (float32 to int32) may lose precision.
+
+This means that it cannot store a ``float32`` result to ``int32``.
+The solution is to type ``ret`` as ``float32`` at the first place:
+
+.. code-block:: python
+
+    @ti.kernel
+    def not_buggy():
+        ret = 0.0  # 0 is a floating point number, so `ret` is typed as float32
+        for i in range(3):
+            ret += 0.1 * i  # f32 += f32, OK!!
+        print(ret)  # will shows 0.6
+
+    not_buggy()
+
+
+`@archibate <https://github.com/archibate>`_'s personal suggestion to prevent issues like this:
+
+* Recall the ``float ret = 0;`` in C/C++, always use ``ret = float(0)`` on **initialization**,
+  and ``ret = int(0)`` for integers. So that you are always clear of what type every variable.
```

Comment:
```suggestion
```

Given the type system can warn the user of this issue, I think they should simply use `ret=0` or `ret=0.0`.